### PR TITLE
Allow multiple default spaces for groups #242

### PR DESCRIPTION
**Pull request by MasterHans**
**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [ ] It's submitted to the `develop` branch, _not_ the `master` branch if no hotfix
- [ ] When resolving a specific issue, it's referenced in the PR's description (e.g. `Fix #xxx[,#xxx]`, where "xxx" is the Github issue number)
- [ ] All tests are passing
- [ ] New/updated tests are included
- [ ] Changelog was modified

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**
1) Allow multiple spaces in the default space field in Group.
2) Create table group_spaces
3) When we saving **GroupUser** we **addMember** to all default Spaces
4) When we saving **GroupSpaces**  also add users of the group to all default spases

Now when we add new space to the default spaces, the system adds the users as a member of all default spaces.

I think we need to move the call of **$group_space->space->addMember($this->user->id);** 
from **GroupUser** **aftersave** event  to **addUser** action and also add  **removeMember** from the spaces to **removeUser** action. But that will be ok if we need to exclude a user from the member of the space when we remove a user from the user Group Member. What is your opinion?

### CODE REVIEW

File: protected/humhub/modules/admin/controllers/GroupController.php [hunk 1]
```diff
@@ -89,6 +89,7 @@ public function actionEdit()
 
         if ($group->load(Yii::$app->request->post()) && $group->validate() && $group->save()) {
             $this->view->saved();
+            Yii::error($group->id);
```
Comment (buddh4) [inline protected/humhub/modules/admin/controllers/GroupController.php:92]: Debug output?

File: protected/humhub/modules/admin/controllers/GroupController.php [hunk 2]
```diff
@@ -262,7 +262,7 @@ public function actionAdminUserSearch($keyword, $id)
     {
         $group = Group::findOne($id);
```
Comment (buddh4) [inline protected/humhub/modules/admin/controllers/GroupController.php:263]: Please change to `findOne(['id' => $id])`

File: protected/humhub/modules/space/models/Space.php [hunk 3]
```diff
@@ -689,4 +690,15 @@ public function getPrivilegedGroupUsers()
 
         return $groups;
     }
+
+    /**
+     * Gets query for [[GroupSpaces]].
+     *
+     * @return \yii\db\ActiveQuery
+     * @since 1.7
```
Comment (buddh4) [inline protected/humhub/modules/space/models/Space.php:698]: 1.8

File: protected/humhub/modules/user/models/Group.php [hunk 4]
```diff
@@ -357,4 +363,13 @@ public static function getDirectoryGroups()
         ])->all();
     }
 
+    /**
+     * Returns all GroupSpaces relations for this group as ActiveQuery.
+     * @return \yii\db\ActiveQuery
+     * @since 1.7
```
Comment (buddh4) [inline protected/humhub/modules/user/models/Group.php:369]: 1.8

File: protected/humhub/modules/user/models/GroupUser.php [hunk 5]
```diff
@@ -82,8 +82,10 @@ public function attributeLabels()
     public function afterSave($insert, $changedAttributes)
     {
         if ($insert) {
-            if ($this->group !== null && $this->group->space !== null) {
-                $this->group->space->addMember($this->user->id);
+            if ($this->group !== null && $this->group->groupSpaces !== null) {
+                foreach ($this->group->groupSpaces as $group_space) {
```
Comment (buddh4) [inline protected/humhub/modules/user/models/GroupUser.php:86]: rather use camel case variable names `groupSpace`

**Review by buddh4 — CHANGES_REQUESTED** [review]

**Review by buddh4 — CHANGES_REQUESTED** [review]
Old default group migration is missing.

File: protected/humhub/modules/user/migrations/m201025_095247_spaces_of_users_group.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+ * Class m201025_095247_spaces_of_users_group
+ */
+class m201025_095247_spaces_of_users_group extends Migration
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function safeUp()
+    {
+        $this->createTable('group_spaces', [
```
Comment (luke-) [inline protected/humhub/modules/user/migrations/m201025_095247_spaces_of_users_group.php:15]: Please do not use a plural here. `group_space`

**Review by luke- — CHANGES_REQUESTED** [review]

File: protected/humhub/modules/admin/controllers/GroupController.php (hunk 1 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/controllers/GroupController.php:92]: Done!

File: protected/humhub/modules/admin/controllers/GroupController.php (hunk 2 above)
Comment (MasterHans) [reply protected/humhub/modules/admin/controllers/GroupController.php:263]: Done!

File: protected/humhub/modules/space/models/Space.php (hunk 3 above)
Comment (MasterHans) [reply protected/humhub/modules/space/models/Space.php:698]: Done!

File: protected/humhub/modules/user/models/Group.php (hunk 4 above)
Comment (MasterHans) [reply protected/humhub/modules/user/models/Group.php:369]: Done!

File: protected/humhub/modules/user/models/GroupUser.php (hunk 5 above)
Comment (MasterHans) [reply protected/humhub/modules/user/models/GroupUser.php:86]: Done!

File: protected/humhub/modules/user/migrations/m201025_095247_spaces_of_users_group.php (hunk 6 above)
Comment (MasterHans) [reply protected/humhub/modules/user/migrations/m201025_095247_spaces_of_users_group.php:15]: Done!

**Comment by MasterHans** [comment]
Guys, please clarify this - should we remove the user from the members of the space if this user was removed from the group member?

**Comment by buddh4** [comment]
@MasterHans no, this is just a default space relation when a user joins the network (or joining a group in the future). Or do you see use cases to sync this relation @luke-. At least I would not do so by default, perhaps by configuration in the future, but in this case we'd also need to sync all existing users when changing the default spaces of a group or assigning a new group to an existing user etc.

**Comment by luke-** [comment]
No, currently it is sufficient if only new users that are added to the group are also added to the Spaces. 

In the second step, possibly as a new PR, I would then introduce a checkbox to reassign existing group members to the default space(s). This would be important in situations where the Spaces change. 

As a third step, we could somehow mark the Space/Group memberships and cancel them if something changes in the group. But here I am not sure if we need it at all.

It is only important that such mass space memberships are realized by an ActiveJob.

**Comment by luke-** [comment]
Here two related issues:

https://github.com/humhub/humhub/issues/4586
https://github.com/humhub/humhub/issues/4585

**Comment by MasterHans** [comment]
@luke-  I have removed automatically reassign of existing group members to the default space(s). For now, only new User Group Members will be assigned to the default space(s).

> Here two related issues:
> 
> #4586
> #4585


I think these new PRs I should perform after this PR will be merged. Am I right?

**Review by luke- — APPROVED** [review]

**Comment by luke-** [comment]
@MasterHans Good question, I think for the next PRs you could use this PR branch als Base Branch

**Comment by MasterHans** [comment]
> Old default group migration is missing.

@buddh4 Done!

**Comment by MasterHans** [comment]
> @MasterHans Good question, I think for the next PRs you could use this PR branch als Base Branch

Ok, I will. Thanks

**Review by luke- — APPROVED** [review]

**Pull request merged**